lib/developer_portal/lib/liquid/tags/cdn_asset.rb

```diff
@@ -48,9 +61,13 @@ def render(context)
 
       private
 
-      def cdn_host
-        hosted_cdn = Rails.configuration.three_scale.assets_cdn_host
-        hosted_cdn ? hosted_cdn : '/_cdn_assets_'
+      def cdn_host(request)
+        rails_config = Rails.configuration.three_scale
+        asset_host_url = rails_config.asset_host.presence || rails_config.assets_cdn_host.presence
```

Comment:
Hmm... but the config `assets_cdn_host` was removed... (`ASSETS_CDN_HOST` value used as a fallback option for `asset_host` config)

Reply:
And I think `rails_config.asset_host.presence` points to a lambda function?

Reply:
`assets_cdn_host` was removed from our `settings.yml` example, but customer's `settings.yml` could still have it. We access it as fallback if `three_scale.asset_host` is not set.

And yes, `rails_config.asset_host.presence` is a [lambda function](https://github.com/3scale/porta/blob/master/config/environments/production.rb#L112-L119) in production.

Reply:
> And yes, rails_config.asset_host.presence is a lambda function in production.

I guess I don't fully understand how this lambda thing works... Like at what stage does it end up being transformed into the actual URL.

Would it ever match `asset_host_url.match? %r{^https?://}` ?

Reply:
I don't think this achieves backward compatibility. If one used `ASSETS_CDN_HOST` in the past, it doesn't mean they used the same for `RAILS_ASSET_HOST` and it doesn't mean that the specified  `ASSETS_CDN_HOST` will be capable of returning any other assets besides the bundled assets for dev portal.

I'm not concerned that users may need to start using another variable to set heir asset host. This is not the backward compatibility issue I'm worried about. Rather I'm worried about users that used a  `ASSETS_CDN_HOST` but a different or no `RAILS_ASSET_HOST`.

It makes little (or no) sense to me for anybody to use such a setup. But we have to ask prod team about it.

Additionally if anybody relied on `ASSETS_CDN_HOST`, they would be serving the assets from the old path `/_cdn_assets_/` rather than the current `/dev-portal-assets/`. So any pre-existing `ASSETS_CDN_HOST` would be invalid either way.

So I don't see any benefit of this commit. My suggestion is to remove this commit. All the rest look fine.

If users had a `RAILS_ASSET_HOST`, then their CDN requirements would be probably still met because that host will be able to cache and serve the built-in portal assets in the same way it does for the rest of the assets. (I assume)

The only group of users that may suffer is if they had  `ASSETS_CDN_HOST` but didn't have `RAILS_ASSET_HOST`. In this case though, if we use that configuration for assets, no assets would be likely to work because of the changed path. But if we ignore the  `ASSETS_CDN_HOST` configuration, then assets will be loaded directly from porta so they should work unless user relied on modifying the assets on their external server.

So IMO if you remove this commit, there would be fewer compatibility issues than if you keep it.

Reply:
> I guess I don't fully understand how this lambda thing works... Like at what stage does it end up being transformed into the actual URL.

Because it's a hell. There are three ways a CDN url can be generated based on what kind of assets is:

- Sprokets:
  - For `javascript_include_tag`, CSSs, images, etc.
  - It prepends `Rails.configuration.asset_host` to the given value
  - `Rails.configuration.asset_host` (at `production.rb`) instead of being just a hostname, is a lambda function that returns `Rails.configuration.three_scale.asset_host` when the given resource actually exists in our assets public folder (don't ask me why). Otherwise it `breaks`, which I don't know the effects of, but I assume the asset loading is cancelled.
-  Webpacker:
    - For `javascript_pack_tag`
        - It uses the same pipeline as Sprockets, so `Rails.configuration.asset_host`.
    - For fonts:
        - CSS styles for fonts are generated on runtime, it prepends `rails_asset_host_url` (at `application_helper.rb`) to the `url()` attribute.
- (New) For developer portal "CDN" assets: It prepends
  - For `cdn_asset` tag in liquid templates.
  - Prepends `cdn_host` (at `cdn_asset.rb`) to the given value.

> Would it ever match asset_host_url.match? %r{^https?://} ?

The lambda function intervenes in a workflow where this never gets executed.

Reply:
Thanks for the explanation @jlledom !

Actually, my confusion was caused by my inability to read properly :clown_face: 
I thought `rails_config` was `Rails.configuration`, while it's actually `Rails.configuration.three_scale` :upside_down_face: 

So, maybe I'd just suggest to rename the variable to `threescale_config`.

Reply:
@mayorova done!

Reply:
@akostadinov 

> Rather I'm worried about users that used a ASSETS_CDN_HOST but a different or no RAILS_ASSET_HOST.

Yeah, this doesn't really work. I think I'm gonna revert it. Any way here's a summary of the behavior before and after this PR, just for reference.

Before:
|RAILS_ASSET_HOST|ASSETS_CDN_HOST|Result|
|---|---|---|
|Empty|Empty| Admin portal assets from `/public`, Developer portal assets from `/cdn` |
|Empty|Set| Admin portal assets from `/public`, Developer portal assets from `ASSETS_CDN_HOST` |
|Set|Empty| Admin portal assets from `RAILS_ASSET_HOST` as cache for `/public`, Developer portal assets from `/cdn`  |
|Set|Set| Admin portal assets from `RAILS_ASSET_HOST` as cache for `/public`, Developer portal assets from `ASSETS_CDN_HOST` |

After:
|RAILS_ASSET_HOST|ASSETS_CDN_HOST|Result|
|---|---|---|
|Empty|Empty| Admin portal assets from `/public`, Developer portal assets from `/public/dev-portal-assets` |
|Empty|Set| Admin portal assets from `/public`, Developer portal assets from `ASSETS_CDN_HOST/dev-portal-assets` as cache  for `/public/dev-portal-assets` |
|Set|Empty| Admin portal assets from `RAILS_ASSET_HOST` as cache for `/public`, Developer portal assets from `RAILS_ASSET_HOST/dev-portal-assets` as cache  for `/public/dev-portal-assets`  |
|Set|Set| Admin portal assets from `RAILS_ASSET_HOST` as cache for `/public`, Developer portal assets from `RAILS_ASSET_HOST/dev-portal-assets` as cache  for `/public/dev-portal-assets` |

And if we revert this commit (`ASSETS_CDN_HOST` would be ignored):
|RAILS_ASSET_HOST|Result|
|---|---|
|Empty| Admin portal assets from `/public`, Developer portal assets from `/public/dev-portal-assets` |
|Set| Admin portal assets from `RAILS_ASSET_HOST` as cache for `/public`, Developer portal assets from `RAILS_ASSET_HOST/dev-portal-assets` as cache  for `/public/dev-portal-assets` 

> Additionally if anybody relied on ASSETS_CDN_HOST, they would be serving the assets from the old path /_cdn_assets_/ rather than the current /dev-portal-assets/. So any pre-existing ASSETS_CDN_HOST would be invalid either way.

This is not exactly right, if anybody relied on `ASSETS_CDN_HOST` they must have a copy of their files in a server on `ASSETS_CDN_HOST`, and `/_cdn_assets_/` would be completely ignored in that case.

Reply:
Yes, I reverted that commit